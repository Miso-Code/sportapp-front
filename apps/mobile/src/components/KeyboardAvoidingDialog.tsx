import React, { ComponentProps, useEffect, useState } from 'react'
import { Dialog, useTheme, MD3Theme } from 'react-native-paper'
import { Keyboard, Platform, StyleSheet } from 'react-native'

const KeyboardAvoidingDialog: React.FC<ComponentProps<typeof Dialog>> = ({
	visible,
	onDismiss,
	children
}) => {
	const [bottom, setBottom] = useState(0)
	const theme = useTheme()
	const styles = createStyles(theme)

	useEffect(() => {
		function onKeyboardChange(e) {
			if (e.endCoordinates.screenY < e.startCoordinates.screenY)
				setBottom(e.endCoordinates.height / 2)
			else setBottom(0)
		}

		if (Platform.OS === 'ios') {
			const subscription = Keyboard.addListener(
				'keyboardWillChangeFrame',
				onKeyboardChange
			)
			return () => subscription.remove()
		}

		const subscriptions = [
			Keyboard.addListener('keyboardDidHide', onKeyboardChange),
			Keyboard.addListener('keyboardDidShow', onKeyboardChange)
		]
		return () =>
			subscriptions.forEach((subscription) => subscription.remove())
	}, [])

	return (
		<Dialog
			testID='dialog'
			style={{ bottom, ...styles.dialog }}
			visible={visible}
			onDismiss={onDismiss}>
			{children}
		</Dialog>
	)
}

const createStyles = (theme: MD3Theme) =>
	StyleSheet.create({
		dialog: {
			backgroundColor: theme.colors.onPrimary,
			borderRadius: 0,
			paddingHorizontal: 20,
			paddingVertical: 10
		}
	})

export default KeyboardAvoidingDialog
