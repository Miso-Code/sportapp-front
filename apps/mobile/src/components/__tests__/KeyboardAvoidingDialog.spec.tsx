import React from 'react'
import renderer, { ReactTestRenderer, act } from 'react-test-renderer'
import { Text } from 'react-native-paper'
import { Keyboard } from 'react-native'

import KeyboardAvoidingDialog from '../KeyboardAvoidingDialog'

jest.mock('react-native-safe-area-context', () => ({
	useSafeAreaInsets: () => ({
		bottom: 0,
		top: 0,
		right: 0,
		left: 0
	}),
	SafeAreaProvider: ({ children }: { children: React.ReactNode }) => children
}))

jest.mock('react-native/Libraries/Utilities/Platform', () => ({
	OS: 'ios',
	select: jest.fn()
}))
jest.mock('react-native/Libraries/Components/Keyboard/Keyboard', () => ({
	addListener: jest.fn().mockReturnValue({ remove: jest.fn() }),
	removeListener: jest.fn()
}))

describe('KeyboardAvoidingDialog', () => {
	let component: ReactTestRenderer

	beforeEach(() => {
		component = renderer.create(
			<KeyboardAvoidingDialog visible={true} onDismiss={jest.fn()}>
				<Text testID='content'>Test</Text>
			</KeyboardAvoidingDialog>
		)
	})

	afterEach(() => {
		component.unmount()
		jest.clearAllMocks()
	})

	it('should render correctly', () => {
		expect(component.toJSON()).toMatchSnapshot()
	})

	it('should render content', () => {
		const [content] = component.root.findAllByProps({ testID: 'content' })
		expect(content.props.children).toBe('Test')
	})

	it('should render dialog', () => {
		const dialog = component.root.findByProps({ testID: 'dialog' })
		expect(dialog.props.visible).toBe(true)
	})

	it('should not render content if not visible', () => {
		act(() => {
			component.update(
				<KeyboardAvoidingDialog visible={false} onDismiss={jest.fn()}>
					<Text testID='content'>Test</Text>
				</KeyboardAvoidingDialog>
			)
		})
		const dialogContentWrapper = component.root.findByProps({
			testID: 'dialog-backdrop'
		})
		expect(
			dialogContentWrapper.props.style[1].opacity._animation?._toValue // since this is an animated value we check the destiny value
		).toBe(0)
	})

	it('should update bottom style prop when keyboard is shown', () => {
		const keyboardEvent = {
			startCoordinates: { screenY: 100, height: 100 },
			endCoordinates: {
				screenY: 0,
				height: 100
			}
		}
		act(() => {
			component.update(
				<KeyboardAvoidingDialog visible={true} onDismiss={jest.fn()}>
					<Text testID='content'>Test</Text>
				</KeyboardAvoidingDialog>
			)
		})
		act(() => {
			;(Keyboard.addListener as jest.Mock).mock.calls[0][1](keyboardEvent)
		})
		const dialog = component.root.findByProps({
			testID: 'dialog'
		})
		expect(dialog.props.style.bottom).toBe(50)
	})

	it('should update bottom style prop to 0 when keyboard is hide', () => {
		const keyboardEvent = {
			startCoordinates: { screenY: 0, height: 100 },
			endCoordinates: {
				screenY: 100,
				height: 100
			}
		}
		act(() => {
			component.update(
				<KeyboardAvoidingDialog visible={true} onDismiss={jest.fn()}>
					<Text testID='content'>Test</Text>
				</KeyboardAvoidingDialog>
			)
		})
		act(() => {
			;(Keyboard.addListener as jest.Mock).mock.calls[0][1](keyboardEvent)
		})
		const dialog = component.root.findByProps({
			testID: 'dialog'
		})
		expect(dialog.props.style.bottom).toBe(0)
	})

	it('should remove all listeners on unmount', () => {
		act(() => {
			component.update(
				<KeyboardAvoidingDialog visible={true} onDismiss={jest.fn()}>
					<Text testID='content'>Test</Text>
				</KeyboardAvoidingDialog>
			)
			component.unmount()
		})
		for (const listener of (Keyboard.addListener as jest.Mock).mock
			.results) {
			expect(listener.value.remove).toHaveBeenCalled()
		}
	})
})
