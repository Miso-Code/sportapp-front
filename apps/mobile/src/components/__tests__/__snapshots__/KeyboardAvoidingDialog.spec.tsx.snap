// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardAvoidingDialog should render correctly 1`] = `
<View
  accessibilityLiveRegion="polite"
  accessibilityViewIsModal={true}
  collapsable={false}
  onAccessibilityEscape={[Function]}
  pointerEvents="auto"
  style={
    {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
  testID="dialog"
>
  <View
    accessibilityLabel="Close modal"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    importantForAccessibility="no"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "rgba(50, 47, 55, 0.4)",
        "flex": 1,
        "opacity": 1,
      }
    }
    testID="dialog-backdrop"
  />
  <View
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "marginBottom": 0,
          "marginTop": 0,
        },
        undefined,
      ]
    }
    testID="dialog-wrapper"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 0,
          "bottom": 0,
          "marginHorizontal": 26,
          "marginVertical": 0,
          "opacity": 1,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 3,
        }
      }
      testID="dialog-surface-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "borderRadius": 0,
            "elevation": 24,
            "flex": undefined,
            "justifyContent": "flex-start",
            "paddingHorizontal": 20,
            "paddingVertical": 10,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 1,
          }
        }
        testID="dialog-surface"
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": undefined,
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "marginTop": 24,
                },
                undefined,
              ],
            ]
          }
          testID="content"
        >
          Test
        </Text>
      </View>
    </View>
  </View>
</View>
`;
